tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,209 @@
+# Stateful tests for arbitrary Zarr stores.
+
+import asyncio
+
+import hypothesis.strategies as st
+from hypothesis import assume, note
+from hypothesis.stateful import (
+    RuleBasedStateMachine,
+    invariant,
+    precondition,
+    rule,
+)
+
+from zarr.core.buffer import Buffer, BufferPrototype, default_buffer_prototype
+from zarr.store import MemoryStore
+from zarr.testing.strategies import key_ranges, paths
+
+
+class SyncStoreWrapper:
+    def __init__(self, store):
+        """Class to hold sync functions that map to async methods of MemoryStore
+        MemoryStore methods are async, this class' methods are sync, so just need to call asyncio.run() in them
+        then, methods in statemachine class are sync and call sync.
+        Unfortunately, hypothesis' stateful testing infra does not support asyncio
+        So we redefine sync versions of the Store API.
+        https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+        """
+        self.store = store
+        self.mode = store.mode
+
+    def set(self, key, data_buffer):
+        return asyncio.run(self.store.set(key, data_buffer))
+
+    def list(self):
+        async def wrapper(gen):
+            return [i async for i in gen]
+
+        gen = self.store.list()
+        yield from asyncio.run(wrapper(gen))
+
+    def get(self, key, prototype: BufferPrototype):
+        obs = asyncio.run(self.store.get(key, prototype=prototype))
+        return obs
+
+    def get_partial_values(self, key_ranges, prototype: BufferPrototype):
+        obs_partial = asyncio.run(
+            self.store.get_partial_values(prototype=prototype, key_ranges=key_ranges)
+        )
+        return obs_partial
+
+    def delete(self, path):
+        return asyncio.run(self.store.delete(path))
+
+    def empty(self):
+        return asyncio.run(self.store.empty())
+
+    def clear(self):
+        return asyncio.run(self.store.clear())
+
+    def exists(self, key):
+        return asyncio.run(self.store.exists(key))
+
+    def list_dir(self, prefix):
+        raise NotImplementedError
+
+    def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    def supports_listing(self):
+        raise NotImplementedError
+
+    def supports_partial_writes(self):
+        raise NotImplementedError
+
+    def supports_writes(self):
+        raise NotImplementedError
```

Comment:
```suggestion
    @property
    def supports_listing(self):
        return self.store.supports_listing

    @property
    def supports_partial_writes(self):
        return self.supports_partial_writes

    @property
    def supports_writes(self):
        return self.store.supports_writes
```

No harm in implementing these.